### DISCUSSION TITLE
Set Multithreaded Compression for sample downloads

### DISCUSSION THREAD

**Discussion by gyto6** [photoprism/photoprism#2324] [Ideas]
Hi, 

I'm using this awesome project to manage my own pictures and share albums. 

The problem is that the users used to select a huge amount of pictures to download. The actual compression is monothreaded and if the compression isn't finished in time, the webpage expires and it's impossible then to download the pictures.

I've a huge server and already optimized everything (14c/28t-128Go RAM - NVMe for Photoprism - SATA SSD for Originals folder - ZFS Dataset optimization including MariaDB - etc...)

I prefered to put in a discussion instead of an issue in case I missed something, but actually, it doesn't sound normal to me not being able  to download too many pictures due to default page expiration and it would be appreciated I think to use most of our cores to speed up the archive generation.

Thank you again

PS: An hint to set multithreaded compression, a bit old. https://www.peterdavehello.org/2015/02/use-multi-threads-to-compress-files-when-taring-something/

**Comment by srett** [comment]
Maybe it makes sense to disable compression, and just store the files in the zip file, i.e. use it as a container only. Should be much faster already. If you mainly have JPGs, you shouldn't see any real reduction in file size anyways with deflate. For the different RAW formats, I'm not so sure what/if the impact of applying compression is, but I think most of them are already compressed, so maybe also neglible.  
Another approach that would still work with compression is streaming the archive to the browser while it's being generated, instead of creating a temp file, if supported by the zip library in use.

**Reply by gyto6** [reply]
Indeed, the pictures doesn't support compression. 

That's not the point of my post, but everywhere I install server, I stopped to use compression within pictures and video's dataset because the benefit is null. 
Dataset are special folder of ZFS for people unaware.
Even on Windows Server with it deduplication, it's not working well with images files. 

"Another approach that would still work with compression is streaming the archive to the browser while it's being generated, instead of creating a temp file, if supported by the zip library in use."

I think that is the solution that should be applied. That's applying when you use the download button on the top right corner, but not for a selection of pictures. The monothreaded side is then no more a problem because the download has begun which is not the case actually. You can't download too many pictures because the request expires before that the archive has been created.

**Reply by lastzero** [reply]
Album downloads are streamed, but not individual files as GET requests are limited in size. Would a streamed download using a POST request work on all devices, including an iOS PWA?

**Reply by gyto6** [reply]
I couldn't answer as I'm not a developer, but I've find this link that might provide some answers ?
https://web.dev/pwa-with-offline-streaming/

**Reply by srett** [reply]
The problem is you can't directly trigger a download on a POST in a PWA/SPA, as it's using `fetch()` / `XMLHttpRequest` which just receives data in your JS.

Either the POST has to store the list of files somewhere in the back-end and then the zip file gets created and streamed in the following GET, which should make it possible to mostly keep the current logic; or you'd need to create an invisible iframe with a `<form>` and POST that, but I think that's rather ugly. But generally speaking (from a traditional Web 1.0 PoV), triggering a download in response to a `POST` is not unusual and supported everywhere.

**Reply by lastzero** [reply]
Note you typically queue zip downloads and don't serve them synchronously as this can easily be exploited for DoS attacks. See Google Drive.

There are many tradeoffs, compatibility, and security considerations to keep in mind. Performance / responsiveness for a single, trusted user as in the problem report is a special use case.
